sdk/core/azure_core/src/policies/bearer_token_policy.rs

```diff
@@ -0,0 +1,50 @@
+use crate::{
+    auth::TokenCredential,
+    headers::AUTHORIZATION,
+    policies::{Policy, PolicyResult},
+    Context, Request,
+};
+use async_trait::async_trait;
+use std::sync::Arc;
+
+#[derive(Debug, Clone)]
+pub struct BearerTokenCredentialPolicy {
+    credential: Arc<dyn TokenCredential>,
+    scopes: Vec<String>,
+}
+
+impl BearerTokenCredentialPolicy {
+    pub fn new<A, B>(credential: Arc<dyn TokenCredential>, scopes: A) -> Self
+    where
+        A: IntoIterator<Item = B>,
+        B: Into<String>,
+    {
+        Self {
+            credential,
+            scopes: scopes.into_iter().map(|s| s.into()).collect(),
+        }
+    }
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+impl Policy for BearerTokenCredentialPolicy {
+    async fn send(
+        &self,
+        ctx: &Context,
+        request: &mut Request,
+        next: &[Arc<dyn Policy>],
+    ) -> PolicyResult {
+        let scopes = self
+            .scopes
+            .iter()
+            .map(|s| s.as_str())
+            .collect::<Vec<&str>>();
+        let access_token = self.credential.get_token(&scopes).await?;
```

Comment:
Does this now seem slightly odd / clunky? The crux of the issue is that `self.credential.get_token` expects type `&[&str]` but since we are storing `scopes` now as `Vec<String>`, there is no zero-cost conversion (to my knowledge) since these iirc look completely different in memory. 

Does this mean `Vec<String>` may not be the best choice here? Another option is to explore whether `get_token` should accept `&[&str]`.

Reply:
You can simplify it slightly: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=71c2cfc4468771da12936a187a4ea5c3

That said, we should probably change `get_token` to be more accepting as well. You can file an issue if you like and think this might be common enough to basically run this routine in the credential impl itself rather than callers.

Reply:
I have taken this feedback- but we can hold off on filing an issue until maybe we see other scenarios that reveal that running this routine in the credential makes more sense than the callers as we do at the moment 😄